apps/src/lab2/views/ExtraLinks.tsx

```diff
@@ -14,61 +19,74 @@ interface PermissionResponse {
-  level_name: string;
-  script_level_path_links: {
-    script: string;
-    path: string;
-  }[];
-}
-
 // If the user has permission to see extra links, fetch extra links for the level,
 // then display a modal with the link data.
 const ExtraLinks: React.FunctionComponent<ExtraLinksProps> = ({
   levelId,
 }: ExtraLinksProps) => {
   const {loading, data} = useFetch('/api/v1/users/current/permissions');
   const [isModalOpen, setIsModalOpen] = useState(false);
-  const [linkData, setLinkData] = useState<ExtraLinksResponse | null>(null);
+  const [levelbuilderLinkData, setLevelbuilderLinkData] =
+    useState<ExtraLinksLevelbuilderData | null>(null);
+  const [projectValidatorLinkData, setProjectValidatorLinkData] =
+    useState<ExtraLinksProjectValidatorData | null>(null);
+  const permissionData = data
+    ? (data as PermissionResponse)
+    : {permissions: []};
+  const channelId = useAppSelector(
+    state => state.lab.channel && state.lab.channel.id
+  );
 
   useEffect(() => {
     const permissionData = data
       ? (data as PermissionResponse)
       : {permissions: []};
-    if (permissionData.permissions.includes(PERMISSIONS.LEVELBUILDER)) {
+    if (
+      permissionData.permissions.includes(PERMISSIONS.LEVELBUILDER) ||
+      permissionData.permissions.includes(PERMISSIONS.PROJECT_VALIDATOR)
+    ) {
       try {
-        HttpClient.fetchJson<ExtraLinksResponse>(
+        HttpClient.fetchJson<ExtraLinksLevelbuilderData>(
           `/levels/${levelId}/extra_links`
         ).then(response => {
-          setLinkData(response.value);
+          setLevelbuilderLinkData(response.value);
         });
       } catch (e) {
-        console.error('Error fetching extra links', e);
+        console.error('Error fetching levelbuilder extra links', e);
       }
     }
-  }, [data, levelId]);
-
-  const permissionData = data
-    ? (data as PermissionResponse)
-    : {permissions: []};
+    if (permissionData.permissions.includes(PERMISSIONS.PROJECT_VALIDATOR)) {
+      try {
+        HttpClient.fetchJson<ExtraLinksProjectValidatorData>(
+          `/projects/${channelId}/extra_links`
+        ).then(response => {
+          setProjectValidatorLinkData(response.value);
+        });
+      } catch (e) {
+        console.error('Error fetching levelbuilder extra links', e);
+      }
+    }
+  }, [data, levelId, channelId]);
 
-  if (!permissionData.permissions.includes(PERMISSIONS.LEVELBUILDER)) {
+  if (
+    !permissionData.permissions.includes(PERMISSIONS.LEVELBUILDER) &&
+    !permissionData.permissions.includes(PERMISSIONS.PROJECT_VALIDATOR)
+  ) {
     return <></>;
   }
 
-  return loading || !linkData ? null : (
+  return loading ||
+    (!levelbuilderLinkData && !projectValidatorLinkData) ? null : (
     <>
       <Button
         onClick={() => setIsModalOpen(true)}
         text={'Extra Links'}
         className={moduleStyles.extraLinksButton}
       />
-      {linkData && (
+      {levelbuilderLinkData && (
```

Comment:
On line 79 above, this will not render unless both level data and link data are defined. I can reset the link data to `null` at the beginning of the `useEffect`. Lmk what you think - thanks!

Reply:
this seems fine then!